napi-inl.h

```diff
@@ -605,6 +606,102 @@ inline Symbol::Symbol(napi_env env, napi_value value) : Name(env, value) {
 }
 
 ////////////////////////////////////////////////////////////////////////////////
+// Automagic value creation
+////////////////////////////////////////////////////////////////////////////////
+
+namespace details {
+template <typename T>
+struct vf_number {
+  static Number From(napi_env env, T value) {
+    return Number::New(env, static_cast<double>(value));
+  }
+};
+
+template<>
+struct vf_number<bool> {
+  static Boolean From(napi_env env, bool value) {
+    return Boolean::New(env, value);
+  }
+};
+
+struct vf_utf8_charp {
+  static String From(napi_env env, const char* value) {
+    return String::New(env, value);
+  }
+};
+
+struct vf_utf16_charp {
+  static String From(napi_env env, const char16_t* value) {
+    return String::New(env, value);
+  }
+};
+struct vf_utf8_string {
+  static String From(napi_env env, const std::string& value) {
+    return String::New(env, value);
+  }
+};
+
+struct vf_utf16_string {
+  static String From(napi_env env, const std::u16string& value) {
+    return String::New(env, value);
+  }
+};
+
+template <typename T>
+struct vf_fallback {
+  static Value From(napi_env env, const T& value) {
+    return Value(env, value);
+  }
+};
+
+template <typename T>
+struct can_make_string {
+  static constexpr bool value =
```

Comment:
There are problems with `constexpr` support on some older compilers. See `NAPI_HAS_CONSTEXPR` elsewhere in this codebase. If this approach doesn't not work at all without `constexpr`, I'm not sure what to do about it.

(Again, we really need a CI system to verify changes work across all supported platforms and versions.)

Reply:
> There are problems with `constexpr` support on some older compilers. See `NAPI_HAS_CONSTEXPR` elsewhere in this codebase. If this approach doesn't not work at all without `constexpr`, I'm not sure what to do about it.

There's definitely a way around this, people did this kind of thing before C++11 :) I am not sure I can do much without access to those older compilers, though, because any attempt at "fixing" this  would be a shot in the dark. I think the obvious approaches would include using `const` instead, using a type instead of a `bool`, or just looking at how `<type_traits>` is implemented by those compilers.

> (Again, we really need a CI system to verify changes work across all supported platforms and versions.)

Yes. :)

Reply:
If all you want to do is avoid this particular constexpr, it can always be rewritten through old-fashioned template metaprogramming. A generic solution would be to implement a disjunction template for testing the `std::is_convertible`s:
```c++
template<class...> struct disjunction : std::false_type { };
template<class B> struct disjunction<B> : B { };
template<class B, class... Bs> struct disjunction<B, Bs...> :
    std::conditional_t<bool(B::value), B, disjunction<Bs...>>  { };
```

Reply:
I made a quick implementation that appears to work: kkoopa@54f5a39ee236b60e82b12d4f3c480e4f48fb0386
```diff
-template <typename T>
 -struct can_make_string {
 -  static constexpr bool value =
 -      std::is_convertible<T, const char*>::value ||
 -      std::is_convertible<T, const char16_t*>::value ||
 -      std::is_convertible<T, std::string>::value ||
 -      std::is_convertible<T, std::u16string>::value;
 -};
 +template <typename...> struct disjunction : std::false_type {};
 +template <typename B> struct disjunction<B> : B {};
 +template <typename B, typename... Bs>
 +struct disjunction<B, Bs...>
 +    : std::conditional<bool(B::value), B, disjunction<Bs...>>::type {};
 +
 +template <typename T>
 +struct can_make_string
 +    : disjunction<typename std::is_convertible<T, const char *>::type,
 +                  typename std::is_convertible<T, const char16_t *>::type,
 +                  typename std::is_convertible<T, std::string>::type,
 +                  typename std::is_convertible<T, std::u16string>::type> {};
```

Reply:
I can't push right now, but if anybody wants to include that into this PR, feel free to do so